packages/nextjs/src/tools/generate-map.ts

```diff
@@ -196,43 +181,118 @@ export const componentMap = new Map<string, NextjsContentSdkComponent>([
 ): string => {
-  return nextjsUnifiedTemplate(components, componentImports, {
-    headerComment:
-      "Below are built-in components that are available in the app, it's recommended to keep them as is",
+  // We don’t pre-process here because this is the legacy “custom template” hook.
+  // This function exists purely as our default client template implementation.
+  const entries = prepareComponentsForMap(components, { includeVariants });
+  return buildNextjsMapContent(entries, componentImports, {
+    headerComment: 'Client-safe component map for App Router',
+    isClientMap: true,
   });
 };
 
-const nextjsMapTemplate = (
-  components: ComponentFile[],
-  componentImports?: ComponentImport[]
-): string => {
-  return nextjsUnifiedTemplate(components, componentImports, {
-    headerComment:
-      "Below are built-in components that are available in the app, it's recommended to keep them as is",
-  });
-};
+/**
+ * Generate and write componentMap.ts files based on provided params.
+ *
+ * When clientComponentMap is true, generates:
+ * - component-map.ts          : Full component map with all components (server, client, universal)
+ * - component-map.client.ts   : Client-safe map with only client + universal components
+ *
+ * When clientComponentMap is false, generates:
+ * - component-map.ts          : Single component map (traditional behavior)
+ *
+ * When includeVariants is true (in either mode):
+ * - Includes component **variants** in the generated map(s) alongside base components
+ * - Preserves the same client/server filtering rules (variants obey clientComponentMap filtering)
+ * - Variant entries are emitted using the same naming/keys convention as their base components
+ *
+ * Template Customization:
+ * - mapTemplate: Custom template for main component map (works for both single and dual map modes)
+ * - clientMapTemplate: Custom template for client component map (only used when clientComponentMap is true)
+ * @param {GenerateMapArgs} param0 params for generateMap
+ */
+export const generateMap: GenerateMapFunction = ({
+  paths,
+  destination = '.sitecore',
+  exclude,
+  componentImports,
+  mapTemplate,
+  clientMapTemplate,
+  clientComponentMap,
+  includeVariants = true,
+}: GenerateMapArgs) => {
+  const isAppRouter = detectRouterType() === 'app';
+  const shouldGenerateClientMap = clientComponentMap ?? isAppRouter;
 
-const nextjsClientMapTemplate = (
-  components: ComponentFileWithType[],
-  componentImports?: ComponentImport[]
-): string => {
-  return nextjsUnifiedTemplate(components, componentImports, {
-    headerComment: 'Client-safe component map for App Router',
-    isClientMap: true,
-  });
+  if (shouldGenerateClientMap) {
+    // App Router case, main map
+    const componentsWithTypes = getComponentListWithTypes(paths, exclude);
+    let mainContent: string;
+    if (mapTemplate) {
+      mainContent = mapTemplate(componentsWithTypes, includeVariants, componentImports);
```

Comment:
should components be prepared in this clause too?
also, maybe the getComponentList/getComponentListWithTypes and prepareComponentsForMap can be bundled together so as to not repeat code and ensure all the required prep steps are always followed